content/graph/LCT.h

```diff
@@ -0,0 +1,116 @@
+/**
+ * Author: 
+ * Description: link-cut Tree. Supports BBST-like 
+ * augmentations. (Can be used in place of HLD).
+ * Current implementation supports update value at a node,
+ * and query max on a path.
+ * Tested on: http://acm.timus.ru/problem.aspx?num=1553
+ * Status: Passes existing fuzz tests (with function names modified).
+ */
+struct Node {
+	bool flip = 0;
+	Node *pp, *p, *c[2];
+	// add stuff
+	int val, cval;
+	Node() { 
+		pp = p = c[0] = c[1] = 0; 
+		val = cval = 0;
+	}
+	// lazy propogation
+	void push() {
+		if (flip) {
+			if(c[0]) c[0]->flip ^= 1;
+			if(c[1]) c[1]->flip ^= 1;
+			swap(c[0], c[1]); flip = 0;
+		}
+		// add stuff
+	}
+	// combine values from children with self.
+	void pull() {
+		cval = val;
+		if(c[0]) c[0]->push(), cval = max(cval, c[0]->cval); 
+		if(c[1]) c[1]->push(), cval = max(cval, c[1]->cval);
+		// add stuff
+	}
+	void rot(bool t) { /// start-hash
+		Node *y = p, *z = y->p;
+		if (z) z->c[z->c[1] == y] = this;
+		p = z;
+		y->c[t] = c[!t];
+		if (c[!t]) c[!t]->p = y;
+		c[!t] = y;
+		y->p = this;
+		if (z) z->pull();
+		y->pull();
```

Comment:
this is an unreadable mess, it wouldn't hurt to golf it a bit and make it even more unreadable... some ideas:
- !t is used more than t, maybe invert its meaning
- `, *&w = y->c[!t]` at the top and use `w` instead of `c[!t]`
- putting things on the same line is possibly worth it, if done tastefully... e.g. maybe `y->p = this` and `p = z` could go together?
- some sort of "fix this node up" function which is called on some subset of c[!t]/this/y/z

Also, I'm a bit surprised the pull for z happens before that of y, I would have expected the other way around.

Reply:
> , *&w = y->c[!t] at the top and use w instead of c[!t]

I might be misunderstanding, but why *&w instead of *w?
Also y->c[!t] is only used once, while this->c[!t] is used multiple times, so *&w = c[!t] might be better.

> some sort of "fix this node up" function which is called on some subset of c[!t]/this/y/z

Like the fix function in the previous lct? (Set a node's children, and then fix will set those children's parents, but without the "(+ update sum of subtree elements etc. if wanted)" part, since that is what pull does).

Pull for z was unnecessary. (A node is only moved up one level for each call of rot, and then pulls up itself at the end of splay.) Thanks for noticing.

Everything else seems good.

Reply:
oh, I confused `y->c[!t]` and `c[!t]`. `c[!t]` is assigned to, hence the `&`.

> Like the fix function in the previous lct?

Yeah. I don't know much about how the previous lct works, but it seems like a method along those lines *may* be able to cut away a few lines? Again, I'm just throwing ideas out, I don't know whether it actually does save anything.

Reply:
Doesn't look like adding fix will shorten it since `w->p = y;` is already as short as `w->fix();`. And then sometimes we'll have to set a parent directly anyway. (when `z = 0` and we need to set `this->p = z` then we need to set `p = z` without calling `z->fix()` anyway). I think it's ok to leave as it is. 
The other changes have been put in.